File: __tests__/lib/contrast.test.ts
// Tests du choix d'encre lisible (lib/contrast.ts).
import { bestInk, contrastRatio, relativeLuminance } from '@/lib/contrast';

const BLANC = '#FFFFFF';
const ENCRE_SOMBRE = '#0B0E13'; // `OnBright` du thème
const MARCHE = '#A3E635'; // teinte lime — la plus claire du thème
const VELO = '#22D3C5';
const MUSCU = '#A78BFA';
const ACCENT = '#5B7CFF';

describe('relativeLuminance', () => {
  it('va de 0 (noir) à 1 (blanc)', () => {
    expect(relativeLuminance('#000000')).toBeCloseTo(0, 5);
    expect(relativeLuminance(BLANC)).toBeCloseTo(1, 5);
  });

  it('accepte la forme courte et ignore le dièse', () => {
    expect(relativeLuminance('#fff')).toBeCloseTo(relativeLuminance(BLANC), 5);
    expect(relativeLuminance('FFFFFF')).toBeCloseTo(relativeLuminance(BLANC), 5);
  });

  it('retombe sur 0 pour une chaîne invalide plutôt que NaN', () => {
    expect(relativeLuminance('pas une couleur')).toBe(0);
    expect(relativeLuminance('#12345')).toBe(0);
  });
});

describe('contrastRatio', () => {
  it('vaut 21 entre noir et blanc, 1 pour une couleur avec elle-même', () => {
    expect(contrastRatio('#000000', BLANC)).toBeCloseTo(21, 2);
    expect(contrastRatio(MARCHE, MARCHE)).toBeCloseTo(1, 5);
  });

  it('est symétrique', () => {
    expect(contrastRatio(MARCHE, BLANC)).toBeCloseTo(contrastRatio(BLANC, MARCHE), 5);
  });
});

describe('bestInk', () => {
  const encres = [ENCRE_SOMBRE, BLANC];

  it('pose une encre sombre sur les teintes claires du thème', () => {
    expect(bestInk(MARCHE, encres)).toBe(ENCRE_SOMBRE);
    expect(bestInk(VELO, encres)).toBe(ENCRE_SOMBRE);
  });

  it('garde le blanc sur un fond sombre', () => {
    expect(bestInk('#000000', encres)).toBe(BLANC);
    expect(bestInk('#14181F', encres)).toBe(BLANC); // surface de carte
  });

  it('retient l’encre sombre sur TOUTES les teintes d’activité du thème', () => {
    // Résultat mesuré, pas choisi : même sur le violet muscu ou le bleu accent,
    // que l'app peignait en blanc, le sombre contraste deux fois mieux.
    for (const teinte of [MARCHE, VELO, MUSCU, ACCENT]) {
      expect(bestInk(teinte, encres)).toBe(ENCRE_SOMBRE);
      expect(contrastRatio(teinte, ENCRE_SOMBRE)).toBeGreaterThan(contrastRatio(teinte, BLANC));
    }
  });

  it('le choix retenu passe le seuil AA sur les teintes du thème', () => {
    for (const teinte of [MARCHE, VELO, MUSCU, ACCENT, '#38BDF8', '#FF5C7A']) {
      expect(contrastRatio(teinte, bestInk(teinte, encres))).toBeGreaterThanOrEqual(4.5);
    }
  });

  it('choisit toujours la meilleure des candidates, même sans seuil atteint', () => {
    // Un gris moyen : aucune encre ne passe AA, on prend quand même la meilleure.
    const gris = '#767676';
    const choisie = bestInk(gris, encres);
    const autre = choisie === BLANC ? ENCRE_SOMBRE : BLANC;
    expect(contrastRatio(gris, choisie)).toBeGreaterThanOrEqual(contrastRatio(gris, autre));
  });
});

File: app/src/test/java/ovh/battistella/elan/domain/ContrastTest.kt
// Tests du choix d'encre lisible (Contrast.kt).
package ovh.battistella.elan.domain

import org.junit.Assert.assertEquals
import org.junit.Assert.assertTrue
import org.junit.Test

class ContrastTest {
    private val blanc = "#FFFFFF"
    private val encreSombre = "#0B0E13" // `OnBright` du thème
    private val marche = "#A3E635" // teinte lime — la plus claire du thème
    private val velo = "#22D3C5"
    private val muscu = "#A78BFA"
    private val accent = "#5B7CFF"
    private val encres = listOf(encreSombre, blanc)

    @Test
    fun `relativeLuminance va de 0 (noir) à 1 (blanc)`() {
        assertEquals(0.0, relativeLuminance("#000000"), 1e-5)
        assertEquals(1.0, relativeLuminance(blanc), 1e-5)
    }

    @Test
    fun `relativeLuminance accepte la forme courte et ignore le dièse`() {
        assertEquals(relativeLuminance(blanc), relativeLuminance("#fff"), 1e-5)
        assertEquals(relativeLuminance(blanc), relativeLuminance("FFFFFF"), 1e-5)
        assertEquals(relativeLuminance(blanc), relativeLuminance("  #ffffff "), 1e-5)
    }

    @Test
    fun `relativeLuminance retombe sur 0 pour une chaîne invalide plutôt que NaN`() {
        assertEquals(0.0, relativeLuminance("pas une couleur"), 0.0)
        assertEquals(0.0, relativeLuminance("#12345"), 0.0)
        assertEquals(0.0, relativeLuminance("#GGGGGG"), 0.0)
    }

    @Test
    fun `contrastRatio vaut 21 entre noir et blanc, 1 pour une couleur avec elle-même`() {
        assertEquals(21.0, contrastRatio("#000000", blanc), 1e-2)
        assertEquals(1.0, contrastRatio(marche, marche), 1e-5)
    }

    @Test
    fun `contrastRatio est symétrique`() {
        assertEquals(contrastRatio(blanc, marche), contrastRatio(marche, blanc), 1e-5)
    }

    @Test
    fun `bestInk pose une encre sombre sur les teintes claires du thème`() {
        assertEquals(encreSombre, bestInk(marche, encres))
        assertEquals(encreSombre, bestInk(velo, encres))
    }

    @Test
    fun `bestInk garde le blanc sur un fond sombre`() {
        assertEquals(blanc, bestInk("#000000", encres))
        assertEquals(blanc, bestInk("#14181F", encres)) // surface de carte
    }

    @Test
    fun `bestInk retient l'encre sombre sur TOUTES les teintes d'activité du thème`() {
        // Résultat mesuré, pas choisi : même sur le violet muscu ou le bleu accent,
        // que l'app peignait en blanc, le sombre contraste deux fois mieux.
        for (teinte in listOf(marche, velo, muscu, accent)) {
            assertEquals(encreSombre, bestInk(teinte, encres))
            assertTrue(contrastRatio(teinte, encreSombre) > contrastRatio(teinte, blanc))
        }
    }

    @Test
    fun `le choix retenu passe le seuil AA sur les teintes du thème`() {
        for (teinte in listOf(marche, velo, muscu, accent, "#38BDF8", "#FF5C7A")) {
            assertTrue(contrastRatio(teinte, bestInk(teinte, encres)) >= 4.5)
        }
    }

    @Test
    fun `bestInk choisit toujours la meilleure des candidates, même sans seuil atteint`() {
        // Un gris moyen : aucune encre ne passe AA, on prend quand même la meilleure.
        val gris = "#767676"
        val choisie = bestInk(gris, encres)
        val autre = if (choisie == blanc) encreSombre else blanc
        assertTrue(contrastRatio(gris, choisie) >= contrastRatio(gris, autre))
    }

    @Test
    fun `bestInk - à égalité, la première candidate l'emporte`() {
        assertEquals("#000000", bestInk("#808080", listOf("#000000", "#000")))
    }
}
